jetty-http/src/main/java/org/eclipse/jetty/http/HttpGenerator.java

```diff
@@ -344,8 +344,19 @@ private Result completing(ByteBuffer chunk, ByteBuffer content)
             _endOfContent = EndOfContent.UNKNOWN_CONTENT;
             return Result.FLUSH;
         }
-
         _state = State.END;
+
+        // If this is an upgrade then we don't want to close the connection.
+        if (_info.isResponse() && ((MetaData.Response)_info).getStatus() == HttpStatus.SWITCHING_PROTOCOLS_101)
+        {
+            return Result.DONE;
+        }
+        else if (_info.isRequest())
+        {
+            HttpField connectionHeader = _info.getFields().getField(HttpHeader.CONNECTION);
+            if (connectionHeader != null && connectionHeader.contains(HttpHeaderValue.UPGRADE.asString()))
+                return Result.DONE;
+        }
```

Comment:
I'm a bit worried about this case.  Just because you send an upgrade doesn't mean that it will be accepted and that we can avoid shutting down the output.   I think we have to ponder this one a bit more.

Reply:
I don't think we really even need this case because our `WebSocketClient` will never send a `Connection: close, Upgrade` header. But it makes the test easier to write because we can use the `WebSocketClient`.

But also because we never send the `close, Upgrade` header this code will never return here, so I don't think it is much of an issue. But I could try re-write the test without the `WebSocketClient` if you want.